### ISSUE TITLE
Issue applying MySQL 5.7.32 with replication Library 3.32

### ISSUE THREAD

**Issue by smithjenne** [mariadb-corporation/mysql-wsrep#381]
I have a 3 node Galera cluster currently running MySQL 5.7.31 on RHEL7.
It is using the wsrep_provider_version 3.31(raf7cd63) .

Today I downloaded the latest binary downloads (mysql-wsrep-5.7.32-25.24-linux-x86_64.tar.gz & galera-3-25.3.32-linux-x86_64.tar.gz).

I followed the same process I normally do:
1) Shutdown one node
2) Unzip the latest binaries (mysql-wsrep-5.7.32-25.24-linux-x86_64.tar.gz & galera-3-25.3.32-linux-x86_64.tar.gz)
3) Restart the node

Here is where I started having problems.  On restart of the first node I got the error:
2021-01-29T16:24:48.017692Z mysqld_safe WSREP: Failed to recover position:  

So, in order to try to isolate the issue I thought I'd try to update the Replication Library (galera-3-25.3.32-linux-x86_64.tar.gz) only.  Well that did work fine.

Then I tried to update (mysql-wsrep-5.7.32-25.24-linux-x86_64.tar.gz) only.  This gave me the "WSREP: Failed to recover position: " error.

There isn't really much more information in the Error log.   Here is everything....
2021-01-29T16:24:45.883240Z mysqld_safe WSREP: Running position recovery with --log_error='/mysql/data/mysql1GC/wsrep_recovery.4r3kMO' --pid-file='/mysql/data/mysql1GC/dba-mdb-d05.andrew.local.cmu.edu-recover.pid'
nohup: ignoring input
2021-01-29T16:24:48.017692Z mysqld_safe WSREP: Failed to recover position: 

Any help in figuring out what is going on would be appreciated.

Thanks,
Jenn

**Comment by smithjenne** [comment]
Hello,
Still having the same issue with the new version on Galera/WSREP 3.33.  Can someone please get back to me on this?

This is now going on 6 months that I cannot upgrade the MySQL environment. 
I'm starting to wonder if anybody looks at these at all.

Thanks,
Jenn

**Comment by temeo** [comment]
Hi,

The log line
```
2021-01-29T16:24:45.883240Z mysqld_safe WSREP: Running position recovery with --log_error='/mysql/data/mysql1GC/wsrep_recovery.4r3kMO' --pid-file='/mysql/data/mysql1GC/dba-mdb-d05.andrew.local.cmu.edu-recover.pid'
```
displays the error log where server log is written during recovery step, in this case `mysql/data/mysql1GC/wsrep_recovery.4r3kMO`. It should contain more information about why the recovery step fails.

**Comment by smithjenne** [comment]
Hi,

So I did look for that file(mysql/data/mysql1GC/wsrep_recovery.4r3kMO), but it doesn't exist.
I did find this file... (/mysql/data/mysql1GC/wsrep_recovery.fail).

However it only contains the following:
2021-05-05T09:43:04.813299-05:00 0 [Warning] TIMESTAMP with implicit DEFAULT value is deprecated. Please use --explicit_defaults_for_timestamp server                                option (see documentation for more details).
2021-05-05T09:43:04.813369-05:00 0 [Warning] 'NO_ZERO_DATE', 'NO_ZERO_IN_DATE' and 'ERROR_FOR_DIVISION_BY_ZERO' sql modes should be used with strict                                mode. They will be merged with strict mode in a future release.
2021-05-05T09:43:04.813374-05:00 0 [Warning] 'NO_AUTO_CREATE_USER' sql mode was not set.
2021-05-05T09:43:04.820324-05:00 0 [Warning] InnoDB: Using innodb_file_format is deprecated and the parameter may be removed in future releases. See                                http://dev.mysql.com/doc/refman/5.7/en/innodb-file-format.html

All of the Warnings listed have been showing up for years.  So nothing new. 

Thanks,
Jenn

**Comment by temeo** [comment]
Only warning messages there... do you happen to have `log_error_verbosity` configured with non default value?

**Comment by smithjenne** [comment]
HI,

Yes I did have log_error_verbosity=2.
unset it in the my.cnf so it is  now the default value of 3.

Here is now what shows up in the log:

2021-05-10T18:40:58.291388Z mysqld_safe WSREP: Failed to recover position:  2021-05-10T14:40:57.874229-05:00 0 
[Warning] TIMESTAMP with implicit DEFAULT value is deprecated. Please use --explicit_defaults_for_timestamp server option (see documentation for more details). 2021-05-10T14:40:57.874304-05:00 0 [Warning] 'NO_ZERO_DATE', 'NO_ZERO_IN_DATE' and 'ERROR_FOR_DIVISION_BY_ZERO' sql modes should be used with strict mode. They will be merged with strict mode in a future release. 2021-05-10T14:40:57.874308-05:00 0
[Warning] 'NO_AUTO_CREATE_USER' sql mode was not set. 2021-05-10T14:40:57.874337-05:00 0 
[Note] --secure-file-priv is set to NULL. Operations related to importing and exporting data are disabled 2021-05-10T14:40:57.874376-05:00 0 
[Note] /mysql/product/MySQLGalera5.7/bin/mysqld (mysqld 5.7.33-log) starting as process 83907 ... 
2021-05-10T14:40:57.881085-05:00 0 
[Warning] InnoDB: Using innodb_file_format is deprecated and the parameter may be removed in future releases. See http://dev.mysql.com/doc/refman/5.7/en/innodb-file-format.html 2021-05-10T14:40:57.881165-05:00 0 [Note] InnoDB: PUNCH HOLE support available 2021-05-10T14:40:57.881193-05:00 0 [Note] InnoDB: Mutexes and rw_locks use GCC atomic builtins 
2021-05-10T14:40:57.881198-05:00 0 [Note] InnoDB: Uses event mutexes 
2021-05-10T14:40:57.881203-05:00 0 [Note] InnoDB: GCC builtin __atomic_thread_fence() is used for memory barrier 
2021-05-10T14:40:57.881208-05:00 0 [Note] InnoDB: Compressed tables use zlib 1.2.11 
2021-05-10T14:40:57.881212-05:00 0 [Note] InnoDB: Using Linux native AIO 
2021-05-10T14:40:57.881985-05:00 0 [Note] InnoDB: Number of pools: 1 
2021-05-10T14:40:57.882127-05:00 0 [Note] InnoDB: Using CPU crc32 instructions 
2021-05-10T14:40:57.884745-05:00 0 [Note] InnoDB: Initializing buffer pool, total size = 128M, instances = 1, chunk size = 128M 
2021-05-10T14:40:57.893608-05:00 0 [Note] InnoDB: Completed initialization of buffer pool 
2021-05-10T14:40:57.896039-05:00 0 [Note] InnoDB: If the mysqld execution user is authorized, page cleaner thread priority can be changed. See the man page of setpriority(). 
2021-05-10T14:40:57.915054-05:00 0 [Note] InnoDB: Highest supported file format is Barracuda. 
2021-05-10T14:40:58.097455-05:00 0 [Note] InnoDB: Creating shared tablespace for temporary tables 
2021-05-10T14:40:58.097506-05:00 0 [Note] InnoDB: Setting file './ibtmp1' size to 12 MB. Physically writing the file full; Please wait ... 
2021-05-10T14:40:58.119852-05:00 0 [Note] InnoDB: File './ibtmp1' size is now 12 MB. 
2021-05-10T14:40:58.120791-05:00 0 [Note] InnoDB: 96 redo rollback segment(s) found. 96 redo rollback segment(s) are active. 
2021-05-10T14:40:58.120804-05:00 0 [Note] InnoDB: 32 non-redo rollback segment(s) are active. 
2021-05-10T14:40:58.121019-05:00 0 [Note] InnoDB: Waiting for purge to start 
2021-05-10T14:40:58.171634-05:00 0 [Note] InnoDB: 5.7.33 started; log sequence number 4628542 
2021-05-10T14:40:58.171918-05:00 0 [Note] Plugin 'FEDERATED' is disabled. 18:40:58 UTC - mysqld got signal 11 ; This could be because you hit a bug. It is also possible that this binary or one of the libraries it was linked against is corrupt, improperly built, or misconfigured. This error can also be caused by malfunctioning hardware. Attempting to collect some information that could help diagnose the problem. As this is a crash and something is definitely wrong, the information collection process might fail. key_buffer_size=8388608 read_buffer_size=131072 max_used_connections=0 max_threads=1000 thread_count=0 connection_count=0 It is possible that mysqld could use up to key_buffer_size + (read_buffer_size + sort_buffer_size)*max_threads = 406137 K bytes of memory Hope that's ok; if not, decrease some variables in the equation. Thread pointer: 0x0 Attempting backtrace. You can use the following information to find out where mysqld died. If you see no messages after this, something went terribly wrong... stack_bottom = 0 thread_stack 0x40000 
/mysql/product/MySQLGalera5.7/bin/mysqld(my_print_stacktrace+0x2c)[0xf3534c] 
/mysql/product/MySQLGalera5.7/bin/mysqld(handle_fatal_signal+0x476)[0x7f6d36] 
/lib64/libpthread.so.0(+0xf630)[0x7f767312a630] 
/mysql/product/MySQLGalera5.7/bin/mysqld[0xebe7de] 
/mysql/product/MySQLGalera5.7/bin/mysqld(_ZN13MYSQL_BIN_LOG10purge_logsEPKcbbbPyb+0x493)[0xecd8a3] 
/mysql/product/MySQLGalera5.7/bin/mysqld(_ZN13MYSQL_BIN_LOG22purge_logs_before_dateElb+0x349)[0xece0a9] 
/mysql/product/MySQLGalera5.7/bin/mysqld[0x7eda26] /mysql/product/MySQLGalera5.7/bin/mysqld(_Z11mysqld_mainiPPc+0xd56)[0x7f0d26] 
/lib64/libc.so.6(__libc_start_main+0xf5)[0x7f7671afe555] 
/mysql/product/MySQLGalera5.7/bin/mysqld[0x7e58d4] 
The manual page at http://dev.mysql.com/doc/mysql/en/crashing.html contains information that should help you find out what is causing the crash.

Thanks,
Jenn

**Comment by sjaakola** [comment]
From the stack trace, it looks that the server crashes when trying to purge binlog files. Binlog files may be corrupted, which caused this.

**Comment by sjaakola** [comment]
@smithjenne please take a look at binlog files in this server, e.g. if you can scan them with mysqlbinlog tool.
If binlog files turn out to be corrupted, you could move them out from this server and try restarting the node again.

**Comment by smithjenne** [comment]
I used mysqlbinlog to scan all binary logs and there were no errors.
Also, if I put the old software back (5.7.31) and do nothing else, I can start the node up cleanly.
If the binlogs were corrupt, wouldn't I see the same issue when starting with the old version?

**Comment by smithjenne** [comment]
Hi,
Based on what you said, I did the following on the first node:
 - Started node1 in 5.7.31
 - purge binary logs to 'master-bin.000055'   <- the latest
 - shutdown node
 - copied 5.7.33 software
 - started node1  <-- Started up fine

However, when I go to the 2nd and 3rd nodes I have the exact same issue.
I shouldn't have to purge the binary logs before doing upgrades right?  I don't mind purging them here cause it's just my playground, but I don't want to have to tell the rest of my team to purge the binary logs on production instances.

Thanks,
Jenn

**Comment by sjaakola** [comment]
Thanks @smithjenne for troubleshooting this. It now turns out that there is a bug, which causes a crash for recovery startup, if there is need for binlog purging at the same time. I will prepare a fix for this issue.
Meanwhile, purging binlog files upfront or preventing binlog file purging by --expire-logs-days configuration should help to circumvent this recovery startup problem.

**Comment by temeo** [comment]
Fixed in 5.7.34-25.26
